diff --git a/linux-libre-deblob b/linux-libre-deblob
index bdee4b5..7ac458b 100644
--- a/linux-libre-deblob
+++ b/linux-libre-deblob
@@ -826,68 +826,6 @@ clean_blob drivers/gpu/drm/display/drm_hdcp_helper.c
 clean_kconfig drivers/gpu/drm/display/Kconfig DRM_DISPLAY_HDCP_HELPER
 clean_mk CONFIG_DRM_DISPLAY_HDCP_HELPER drivers/gpu/drm/display/Makefile

-announce DRM_AMDGPU - "AMD GPU"
-reject_firmware drivers/gpu/drm/amd/amdgpu/amdgpu_uvd.c
-clean_blob drivers/gpu/drm/amd/amdgpu/amdgpu_uvd.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/amdgpu_vce.c
-clean_blob drivers/gpu/drm/amd/amdgpu/amdgpu_vce.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/amdgpu_vcn.c
-clean_blob drivers/gpu/drm/amd/amdgpu/amdgpu_vcn.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/gfx_v8_0.c
-clean_blob drivers/gpu/drm/amd/amdgpu/gfx_v8_0.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/gfx_v9_0.c
-clean_blob drivers/gpu/drm/amd/amdgpu/gfx_v9_0.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/gfx_v10_0.c
-clean_blob drivers/gpu/drm/amd/amdgpu/gfx_v10_0.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/gfx_v11_0.c
-clean_blob drivers/gpu/drm/amd/amdgpu/gfx_v11_0.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/gmc_v8_0.c
-clean_blob drivers/gpu/drm/amd/amdgpu/gmc_v8_0.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/imu_v11_0.c
-clean_blob drivers/gpu/drm/amd/amdgpu/imu_v11_0.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/mes_v10_1.c
-clean_blob drivers/gpu/drm/amd/amdgpu/mes_v10_1.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/mes_v11_0.c
-clean_blob drivers/gpu/drm/amd/amdgpu/mes_v11_0.c
-clean_blob drivers/gpu/drm/amd/amdgpu/psp_v3_1.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/psp_v10_0.c
-clean_blob drivers/gpu/drm/amd/amdgpu/psp_v10_0.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/psp_v11_0.c
-clean_blob drivers/gpu/drm/amd/amdgpu/psp_v11_0.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/psp_v12_0.c
-clean_blob drivers/gpu/drm/amd/amdgpu/psp_v12_0.c
-clean_blob drivers/gpu/drm/amd/amdgpu/psp_v13_0.c
-clean_blob drivers/gpu/drm/amd/amdgpu/psp_v13_0_4.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/sdma_v2_4.c
-clean_blob drivers/gpu/drm/amd/amdgpu/sdma_v2_4.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/sdma_v3_0.c
-clean_blob drivers/gpu/drm/amd/amdgpu/sdma_v3_0.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/sdma_v4_0.c
-clean_blob drivers/gpu/drm/amd/amdgpu/sdma_v4_0.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/sdma_v5_0.c
-clean_blob drivers/gpu/drm/amd/amdgpu/sdma_v5_0.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/sdma_v5_2.c
-clean_blob drivers/gpu/drm/amd/amdgpu/sdma_v5_2.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/sdma_v6_0.c
-clean_blob drivers/gpu/drm/amd/amdgpu/sdma_v6_0.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/amdgpu_cgs.c
-clean_blob drivers/gpu/drm/amd/amdgpu/amdgpu_cgs.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/amdgpu_device.c
-clean_blob drivers/gpu/drm/amd/amdgpu/amdgpu_device.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/amdgpu_psp.c
-clean_blob drivers/gpu/drm/amd/amdgpu/amdgpu_psp.c
-clean_blob drivers/gpu/drm/amd/pm/powerplay/smumgr/smumgr.c
-reject_firmware drivers/gpu/drm/amd/pm/swsmu/smu11/smu_v11_0.c
-clean_blob drivers/gpu/drm/amd/pm/swsmu/smu11/smu_v11_0.c
-reject_firmware drivers/gpu/drm/amd/pm/swsmu/smu13/smu_v13_0.c
-clean_blob drivers/gpu/drm/amd/pm/swsmu/smu13/smu_v13_0.c
-reject_firmware drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c
-clean_blob drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c
-reject_firmware drivers/gpu/drm/amd/amdgpu/amdgpu_discovery.c
-clean_blob drivers/gpu/drm/amd/amdgpu/amdgpu_discovery.c
-clean_kconfig drivers/gpu/drm/Kconfig DRM_AMDGPU
-clean_mk CONFIG_DRM_AMDGPU drivers/gpu/drm/amd/amdgpu/Makefile
-
 announce DRM_AMDGPU_CIK - "Enable amdgpu support for CIK parts"
 reject_firmware drivers/gpu/drm/amd/amdgpu/cik_sdma.c
 clean_blob drivers/gpu/drm/amd/amdgpu/cik_sdma.c
@@ -2145,16 +2083,6 @@ clean_blob drivers/net/wireless/intel/iwlegacy/4965.c
 clean_kconfig drivers/net/wireless/intel/iwlegacy/Kconfig IWL4965
 clean_mk CONFIG_IWL4965 drivers/net/wireless/intel/iwlegacy/Makefile

-announce IWLWIFI - "Intel Wireless WiFi Next Gen AGN"
-reject_firmware drivers/net/wireless/intel/iwlwifi/iwl-nvm-parse.c
-reject_firmware drivers/net/wireless/intel/iwlwifi/iwl-drv.c
-reject_firmware drivers/net/wireless/intel/iwlwifi/iwl-dbg-tlv.c
-reject_firmware drivers/net/wireless/intel/iwlwifi/fw/pnvm.c
-clean_blob drivers/net/wireless/intel/iwlwifi/iwl-drv.c
-clean_blob drivers/net/wireless/intel/iwlwifi/iwl-dbg-tlv.c
-clean_kconfig drivers/net/wireless/intel/iwlwifi/Kconfig IWLWIFI
-clean_mk CONFIG_IWLWIFI drivers/net/wireless/intel/iwlwifi/Makefile
-
 announce IWLDVM - "Intel Wireless WiFi DVM Firmware support"
 clean_blob drivers/net/wireless/intel/iwlwifi/cfg/1000.c
 clean_blob drivers/net/wireless/intel/iwlwifi/cfg/2000.c
@@ -2163,14 +2091,6 @@ clean_blob drivers/net/wireless/intel/iwlwifi/cfg/6000.c
 clean_kconfig drivers/net/wireless/intel/iwlwifi/Kconfig IWLDVM
 clean_mk CONFIG_IWLMVM drivers/net/wireless/intel/iwlwifi/Makefile

-announce IWLMVM - "Intel Wireless WiFi MVM Firmware support"
-clean_blob drivers/net/wireless/intel/iwlwifi/cfg/7000.c
-clean_blob drivers/net/wireless/intel/iwlwifi/cfg/8000.c
-clean_blob drivers/net/wireless/intel/iwlwifi/cfg/9000.c
-clean_blob drivers/net/wireless/intel/iwlwifi/cfg/22000.c
-clean_kconfig drivers/net/wireless/intel/iwlwifi/Kconfig IWLMVM
-clean_mk CONFIG_IWLMVM drivers/net/wireless/intel/iwlwifi/Makefile
-
 announce KS7010 - "KeyStream KS7010 SDIO support"
 reject_firmware drivers/staging/ks7010/ks7010_sdio.c
 clean_blob drivers/staging/ks7010/ks7010_sdio.c
