diff --git a/build.rs b/build.rs
index ea34c22..4bffe86 100644
--- a/build.rs
+++ b/build.rs
@@ -32,7 +32,7 @@ fn main() {
         // can be called via FFI
         .wrap_static_fns(true)
         .wrap_static_fns_suffix("_libspa_rs")
-        .wrap_static_fns_path(&out_path.join("static_fns"));
+        .wrap_static_fns_path(out_path.join("static_fns"));
 
     let builder = libs
         .iter()
diff --git a/src/type_info.rs b/src/type_info.rs
index 016c971..e1e60e9 100644
--- a/src/type_info.rs
+++ b/src/type_info.rs
@@ -101,9 +101,7 @@ mod test {
             let type_info = super::spa_debug_type_find(spa_type_media_type, SPA_MEDIA_TYPE_audio);
             assert_eq!(
                 ffi::CStr::from_ptr((*type_info).name),
-                ffi::CString::new("Spa:Enum:MediaType:audio")
-                    .unwrap()
-                    .as_ref()
+                c"Spa:Enum:MediaType:audio"
             );
         }
     }
@@ -112,12 +110,7 @@ mod test {
     fn test_libspa_rs_debug_type_find_name() {
         unsafe {
             let name = super::spa_debug_type_find_name(spa_type_media_type, SPA_MEDIA_TYPE_audio);
-            assert_eq!(
-                ffi::CStr::from_ptr(name),
-                ffi::CString::new("Spa:Enum:MediaType:audio")
-                    .unwrap()
-                    .as_ref()
-            );
+            assert_eq!(ffi::CStr::from_ptr(name), c"Spa:Enum:MediaType:audio");
         }
     }
 
@@ -126,10 +119,7 @@ mod test {
         unsafe {
             let name =
                 super::spa_debug_type_find_short_name(spa_type_media_type, SPA_MEDIA_TYPE_audio);
-            assert_eq!(
-                ffi::CStr::from_ptr(name),
-                ffi::CString::new("audio").unwrap().as_ref()
-            );
+            assert_eq!(ffi::CStr::from_ptr(name), c"audio");
         }
     }
 }
